### ISSUE TITLE
Calculations Testing

### ISSUE THREAD

**Issue by ravenwilde** [kisanjong/rotalabutterfly#9]
Please be as specific as possible when reporting Calculations errors. It is most helpful for us if you list the following:
- The compound or premixed product that you tested
- All other form fields that you had filled in

Thank you :)

**Comment by kisanjong** [comment]
I've posted to Barr report, ukaps and planted tank requesting volunteers for beta testing.

**Comment by fablau** [comment]
Hey guys, I found something missing in the latest version:

http://rotalabutterfly.com/test/

If you want to find a solution of CSM+B which exceed the solubility of the compound, you don't get ay warning. Whereas my PHP backend is returning the "sol_error" flag, I can't see that shown in the new JS implementation. Jennifer, thoughts?

Here is what I tried:

75gl tank, DIY, CSM+B, a solution, 500ml container, 5ml dose (too little for solubility)

**Comment by kisanjong** [comment]
I will post comments as i receive them from testers.

"I tested for KNO3 for EI dosing a DIY solution. 

However, the result only told me how many grams of dry ferts to add to my aquarium. 

What I expected from the old version is that it should tell me how many grams/teaspoons to add to the water in the dosing container"

**Comment by kisanjong** [comment]
"Can we change the contact link to contact & feedback"

**Comment by kisanjong** [comment]
"I just tested the Test page and no output is provided for the concentrations."

**Comment by kisanjong** [comment]
Two comments now mentioned grams/teaspoons not just grams

**Comment by kisanjong** [comment]
Fab I've just tested your csm+b and get this

![screenshot_2015-07-22-22-28-55](https://cloud.githubusercontent.com/assets/12550864/8837991/5fd8fa12-30c1-11e5-8f09-beaf56675697.png)

**Comment by fablau** [comment]
Yes Jason, I get the same results, and they should be correct. What I meant is that solubility for that dosing (5ml) would have some issues, and the calculator should warn that. The PHP backend passes the variable  "sol_error" set to 1 in case solubility issues are reached. I think the UI is just not handling that passed parameter. Does this make sense?

**Comment by fablau** [comment]
> Two comments now mentioned grams/teaspoons not just grams

I can add that calculation to the PHP backend, and return the conversion in tsp or tbsp as an additional variable.

**Comment by ravenwilde** [comment]
> "Can we change the contact link to contact & feedback"

I'd like to steer away from this - its overlong for a navigation link. What we could do instead is display a note below the calculator that says something along the lines of "This is an evolving project that is currently in beta, your feedback is appreciated, please use our [link]contact form[/link].

We could display the same/a similar message on the contact page itself.

Acceptable solution?

**Comment by kisanjong** [comment]
Sounds good to me Jennifer :)

**Comment by kisanjong** [comment]
"Also labeling urea as toxic is misleading. It's only toxic at a certain concentration like all of the elements."

**Comment by kisanjong** [comment]
Fab it makes sense now and sounds like a good idea to warn about this

**Comment by Audionut** [comment]
Would be nice to be able to choose the target element when calculating for "dose to reach target".

For instance, if I wanted to raise dGH by ???

edit:  After the calculator is up and running, if you wanted to get really tricky, you could do something like.
"I want to target ??? dGH, using XXX and XXX at a Y:Y ratio"

**Comment by kisanjong** [comment]
Some more feedback from testers...

So, some general feedback/wishlist here after tinkering with it for a bit, some of which is inherited from Wet's original calculator... Thus far looks like a good start.  Issues specific to the test calculator, most of which is probably due to the beta nature and may already be well known to you:  

1) Computing "Result of my dose" doesn't seem to work yet... (this is a beta, so not too surprising) 

 2) Computing "Dose to reach target" doesn't allow non-integer ppm's... I often target in 0.1 ppm increments for things like Iron, and 0.5 increments for PO4.  

 3) Computing EI for Millers microplex results in a rather cryptic toxicity warning:  "toxic    cu,163,Miller's MicroPlex,1.06,grams"  I'm guessing that means Cu is toxic at 0.163ppm, which would be provided by 1.06 grams of Microplex in my 36 gallon tank, but the dose to reach EI Iron levels is 1.70 grams. (which results in 0.187ppm Cu. 

 Wishlist things that were issues in Wet's calculator:  

1) It has always bothered me that micronutrient mixes like millers and CSM+B are only regarded as iron supplements when being used to calculate dosing. I'd love to be able to do things like calculate "dose to reach a target" on nutrients other than just iron. So a picker that lets you change the target nutrient would be helpful.  

2) I'd love to have something like barr's gh booster added to DIY, particularly if you could target it based on dGH... (3 parts K2SO4, 3 parts CaSO4, 1 part MgSO4).  3) Calculations targeting MgSO4/CaSO4 to ppm of Mg/Ca or dGH would also be nice.

**Comment by ravenwilde** [comment]
> If you want to find a solution of CSM+B which exceed the solubility of the compound, you don't get ay warning. Whereas my PHP backend is returning the "sol_error" flag, I can't see that shown in the new JS implementation. Jennifer, thoughts?

I'll make sure I take care of this when I begin working on error messages/form validation, etc. I created an Issue so I won't forget.

**Comment by kisanjong** [comment]
Thanks :), one comment that keeps popping up is the toxicity warnings

**Comment by ravenwilde** [comment]
I'm working on the language selection right now, I'll do the error flags and toxicity warnings next.

**Comment by kisanjong** [comment]
Nice :D

Do we have any Italians that can test this? ;)

**Comment by ravenwilde** [comment]
We're going to need all the languages tested - Because the form labels on our calculator are slightly different than Wet's original I had to adjust the translations - I used Google translate to do so and they will definitely need double checking. I will put a note on the interface asking for help with this.

**Comment by fablau** [comment]
> Do we have any Italians that can test this? ;)

I am Italian guys! What do you want me to test?

**Comment by fablau** [comment]
Jason, I think we can do pretty much everything you have described above... step-by-step we can do it! Jennifer: just let me know when you need me to step-in. Thanks :)

**Comment by fablau** [comment]
> Fab it makes sense now and sounds like a good idea to warn about this

That was taken from original Wet's implementation. We can change whatever we like. As you already said, this is a good start, but we can shape it as we like.

**Comment by kisanjong** [comment]
Hey Fab I know you are Italian hence my wink smile :) 

It's all sounding great and the future addons are endless.

Thankyou both :D

**Comment by fablau** [comment]
> Hey Fab I know you are Italian hence my wink smile :)
> It's all sounding great and the future addons are endless.
> Thankyou both :D

Sorry, I missed that!! ;) Just woke up, still sleepy! :)
